In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        21:10:42   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Q4: R-squared (17.6% Variability Explained): The R-squared value of 17.6% indicates that only a small portion of the variance in the dependent variable (HP) is explained by the predictors (Sp. Def, Generation, and their interaction terms) in your model. This implies that there are likely other factors influencing HP that are not included in this model, meaning it is not a comprehensive or fully predictive model of HP.

Large Coefficients with Strong Statistical Significance: The large coefficients and strong statistical evidence against the null hypothesis for some terms suggest that, within the context of this model, changes in Sp. Def, Generation, or the interactions are indeed associated with changes in HP. However, this association does not necessarily imply a strong explanatory power across the full range of variability in HP — it only indicates that there is a detectable, statistically significant relationship within the confines of what this model assesses.

Q7: 
From Model 3,4 to Model 5: Model 5 further expands on this by adding additional predictors and some categorical variables to more thoroughly capture complex patterns in HP. This increases explanatory power by incorporating more potentially relevant characteristics without overwhelming interaction terms.
From Model 5 to Model 6: The goal here is parsimony—retaining only those predictors with demonstrated importance. This simplifies the model while maintaining predictive accuracy.
From Model 6 to Model 7: Scaling and centering are applied to numeric variables to improve stability, particularly helpful in a model with high interactions, thereby reducing multicollinearity and enhancing robustness.

Q9: The code examines how well the regression models predict Pokémon HP for different subsets of the data based on generation. By splitting the data into groups (e.g., Generation 1 vs. Generation 6), we can see if the model trained on one group performs well when applied to a different generation. This is important because it helps us understand whether the model can generalize to unseen Pokémon or if it's overfitting the data it was trained on. High in-sample R-squared values mean a good fit, but the real test is whether the model can predict accurately for different generations of Pokémon, which is reflected in the out-of-sample R-squared.

Chatbot history link: https://chatgpt.com/share/67367824-f344-800a-bc4f-dc48b6067749